pandas/util/_decorators.py

```diff
@@ -268,17 +269,26 @@ def decorator(func: F) -> F:
         def wrapper(*args, **kwargs) -> Callable:
             return func(*args, **kwargs)
 
-        templates = [func.__doc__ if func.__doc__ else ""]
+        # collecting and docstring templates
+        wrapper._doc_args: List[Union[str, Callable]] = []  # type: ignore
+        if func.__doc__:
+            wrapper._doc_args.append(dedent(func.__doc__))  # type: ignore
+
         for arg in args:
-            if isinstance(arg, str):
-                templates.append(arg)
-            elif hasattr(arg, "_docstr_template"):
-                templates.append(arg._docstr_template)  # type: ignore
-            elif arg.__doc__:
-                templates.append(arg.__doc__)
-
-        wrapper._docstr_template = "".join(dedent(t) for t in templates)  # type: ignore
-        wrapper.__doc__ = wrapper._docstr_template.format(**kwargs)  # type: ignore
+            if hasattr(arg, "_doc_args"):
+                wrapper._doc_args.extend(arg._doc_args)  # type: ignore
+            elif isinstance(arg, str) or arg.__doc__:
+                wrapper._doc_args.append(arg)  # type: ignore
+
+        # formatting templates and concatenating docstring
+        wrapper.__doc__ = "".join(
+            [
+                arg.format(**kwargs)
+                if isinstance(arg, str)
+                else dedent(arg.__doc__)  # type: ignore
+                for arg in wrapper._doc_args  # type: ignore
```

Comment:
can you confirm that these `# type: ignore` stem from the fact that we are adding attributes to the wrapper function?

I guess this was discussed originally, but would using a class for the doc decorator help?

Reply:
Good point here! Most `# type: ignore` is because we are adding attributes to function. 

The only one not for this reason is `else dedent(arg.__doc__)  # type: ignore` (line 288). This one is because theoretically, `arg.__doc__` can be `str` or `None`. However, we have checked `arg.__doc__` is not `None` before putting it in, so that we are safe here.

Back to the most `# type: ignore` cases. I strongly agree with you that we could consider solving this issue without using `# type: ignore`, because this looks like an abuse. I would like to try putting it under a class. Thanks for this advice!

Reply:
> The only one not for this reason is `else dedent(arg.__doc__) # type: ignore` (line 288). This one is because theoretically, `arg.__doc__` can be `str` or `None`. However, we have checked `arg.__doc__` is not `None` before putting it in, so that we are safe here.

i _think_ ok to remove
```
$ mypy pandas --warn-unused-ignores
pandas\util\_decorators.py:288: error: unused 'type: ignore' comment
```

Reply:
@simonjayhawkins Good catch here! Thanks. [Fixing](https://github.com/pandas-dev/pandas/pull/31970/commits/7bf242e03bc10aeee4a7c6df515df23fbca23859) is here.